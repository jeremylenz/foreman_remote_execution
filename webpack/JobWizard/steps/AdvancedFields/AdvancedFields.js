import React from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { Title, Form } from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';
import {
  selectEffectiveUser,
  selectAdvancedTemplateInputs,
} from '../../JobWizardSelectors';
import {
  EffectiveUserField,
  TimeoutToKillField,
  PasswordField,
  KeyPassphraseField,
  EffectiveUserPasswordField,
  ConcurrencyLevelField,
  TimeSpanLevelField,
  TemplateInputsFields,
} from './Fields';

export const AdvancedFields = ({ advancedValues, setAdvancedValues }) => {
  const effectiveUser = useSelector(selectEffectiveUser);
  const templateInputs = useSelector(selectAdvancedTemplateInputs);
  return (
    <>
      <Title headingLevel="h2" className="advanced-fields-title">
        {__('Advanced Fields')}
      </Title>
      <Form id="advanced-fields-job-template" autoComplete="off">
        <TemplateInputsFields
          inputs={templateInputs}
          value={advancedValues.templateValues}
          setValue={newValue => setAdvancedValues({ templateValues: newValue })}
        />
        {effectiveUser?.overridable && (
          <EffectiveUserField
            value={advancedValues.effectiveUserValue}
            setValue={newValue =>
              setAdvancedValues({
                effectiveUserValue: newValue,
              })
            }
          />
        )}
        <TimeoutToKillField
          value={advancedValues.timeoutToKill}
          setValue={newValue =>
            setAdvancedValues({
              timeoutToKill: newValue,
            })
          }
        />
        <PasswordField
          value={advancedValues.password}
          setValue={newValue =>
            setAdvancedValues({
              password: newValue,
            })
          }
        />
        <KeyPassphraseField
          value={advancedValues.keyPassphrase}
          setValue={newValue =>
            setAdvancedValues({
              keyPassphrase: newValue,
            })
          }
        />
        <EffectiveUserPasswordField
          value={advancedValues.effectiveUserPassword}
          setValue={newValue =>
            setAdvancedValues({
              effectiveUserPassword: newValue,
            })
          }
        />
        <ConcurrencyLevelField
          value={advancedValues.concurrencyLevel}
          setValue={newValue =>
            setAdvancedValues({
              concurrencyLevel: newValue,
            })
          }
        />
        <TimeSpanLevelField
          value={advancedValues.timeSpan}
          setValue={newValue =>
            setAdvancedValues({
              timeSpan: newValue,
            })
          }
        />
      </Form>
    </>
  );
};

AdvancedFields.propTypes = {
  advancedValues: PropTypes.object.isRequired,
  setAdvancedValues: PropTypes.func.isRequired,
};
export default AdvancedFields;
