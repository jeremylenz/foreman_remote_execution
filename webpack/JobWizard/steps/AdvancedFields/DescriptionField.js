import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { FormGroup, TextInput, Button } from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';

export const DescriptionField = ({ inputs, value, setValue }) => {
  const generateDesc = () => {
    let newDesc = value;
    if (value) {
      const re = new RegExp('%\\{([^\\}]+)\\}', 'gm');
      const results = [...newDesc.matchAll(re)].map(result => ({
        name: result[1],
        text: result[0],
      }));
      results.forEach(result => {
        newDesc = newDesc.replace(
          result.text,
          // TODO: Replace with the value of the input from Target Hosts step
          inputs.find(input => input.name === result.name)?.name || result.text
        );
      });
    }
    return newDesc;
  };
  const [generatedDesc, setGeneratedDesc] = useState(generateDesc());
  const [isPreview, setIsPreview] = useState(true);

  const togglePreview = () => {
    setGeneratedDesc(generateDesc());
    setIsPreview(v => !v);
  };

  return (
    <FormGroup
      label={__('Description')}
      fieldId="description"
      helperText={
        <Button variant="link" isInline onClick={togglePreview}>
          {isPreview
            ? __('Edit job description template')
            : __('Preview job description')}
        </Button>
      }
    >
      {isPreview ? (
        <TextInput id="description-preview" value={generatedDesc} isDisabled />
      ) : (
        <TextInput
          type="text"
          autoComplete="description"
          id="description"
          value={value}
          onChange={newValue => setValue(newValue)}
        />
      )}
    </FormGroup>
  );
};

DescriptionField.propTypes = {
  inputs: PropTypes.array.isRequired,
  value: PropTypes.string,
  setValue: PropTypes.func.isRequired,
};
DescriptionField.defaultProps = {
  value: '',
};
