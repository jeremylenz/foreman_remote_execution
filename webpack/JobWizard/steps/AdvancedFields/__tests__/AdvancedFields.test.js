import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import * as patternfly from '@patternfly/react-core';
import { mount } from '@theforeman/test';
import { AdvancedFields } from '../AdvancedFields';
import { jobTemplateResponse } from '../../../__tests__/fixtures';

jest.spyOn(patternfly, 'FormGroup');
patternfly.FormGroup.mockImplementation(props => (
  <div>{props.navAriaLabel}</div>
));
const mockStore = configureMockStore([]);
const store = mockStore({
  JOB_TEMPLATE: { response: jobTemplateResponse },
});
describe('AdvancedFields', () => {
  it('rendring', () => {
    const component = mount(
      <Provider store={store}>
        <AdvancedFields advancedValues={{}} setAdvancedValues={jest.fn()} />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
