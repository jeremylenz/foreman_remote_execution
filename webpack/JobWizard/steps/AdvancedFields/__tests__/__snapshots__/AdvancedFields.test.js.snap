// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedFields rendring 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AdvancedFields
    advancedValues={Object {}}
    setAdvancedValues={[MockFunction]}
  >
    <Title
      className="advanced-fields-title"
      headingLevel="h2"
    >
      <h2
        className="pf-c-title pf-m-xl advanced-fields-title"
      >
        Advanced Fields
      </h2>
    </Title>
    <Form
      autoComplete="off"
      id="advanced-fields-job-template"
    >
      <form
        autoComplete="off"
        className="pf-c-form"
        id="advanced-fields-job-template"
        noValidate={true}
      >
        <TemplateInputsFields
          inputs={
            Array [
              Object {
                "advanced": true,
                "default": "Default val",
                "description": "some Description",
                "hidden_value": true,
                "input_type": "user",
                "name": "adv plain hidden",
                "required": true,
                "resource_type": "ansible_roles",
                "value_type": "plain",
              },
            ]
          }
          setValue={[Function]}
          value={Object {}}
        >
          <mockConstructor
            fieldId="adv plain hidden"
            isRequired={true}
            key="adv plain hidden"
            label="adv plain hidden"
            labelIcon={
              <Popover
                aria-label="help-text"
                bodyContent="some Description"
                id="adv plain hidden-help"
              >
                <button
                  aria-label="open-help-tooltip-button"
                  className="pf-c-form__group-label-help"
                  onClick={[Function]}
                  type="button"
                >
                  <HelpIcon
                    color="currentColor"
                    noVerticalAlign={true}
                    size="sm"
                  />
                </button>
              </Popover>
            }
          >
            <div />
          </mockConstructor>
        </TemplateInputsFields>
        <EffectiveUserField
          setValue={[Function]}
          value=""
        >
          <mockConstructor
            fieldId="effective-user"
            label="Effective user"
            labelIcon={
              <Popover
                aria-label="help-text"
                bodyContent="A user to be used for executing the script. If it differs from the SSH user, su or sudo is used to switch the accounts."
                id="effective-user-help"
              >
                <button
                  aria-label="open-help-tooltip-button"
                  className="pf-c-form__group-label-help"
                  onClick={[Function]}
                  type="button"
                >
                  <HelpIcon
                    color="currentColor"
                    noVerticalAlign={true}
                    size="sm"
                  />
                </button>
              </Popover>
            }
          >
            <div />
          </mockConstructor>
        </EffectiveUserField>
        <TimeoutToKillField
          setValue={[Function]}
          value=""
        >
          <NumberInput
            formProps={
              Object {
                "fieldId": "timeout-to-kill",
                "label": "Timeout to kill",
                "labelIcon": <Popover
                  aria-label="help-text"
                  bodyContent="Time in seconds from the start on the remote host after which the job should be killed."
                  id="timeout-to-kill-help"
                >
                  <button
                    aria-label="open-help-tooltip-button"
                    className="pf-c-form__group-label-help"
                    onClick={[Function]}
                    type="button"
                  >
                    <HelpIcon
                      color="currentColor"
                      noVerticalAlign={true}
                      size="sm"
                    />
                  </button>
                </Popover>,
              }
            }
            inputProps={
              Object {
                "autoComplete": "timeout-to-kill",
                "id": "timeout-to-kill",
                "onChange": [Function],
                "placeholder": "For example: 1, 2, 3, 4, 5...",
                "value": "",
              }
            }
          >
            <mockConstructor
              fieldId="timeout-to-kill"
              helperTextInvalid="Has to be a number"
              label="Timeout to kill"
              labelIcon={
                <Popover
                  aria-label="help-text"
                  bodyContent="Time in seconds from the start on the remote host after which the job should be killed."
                  id="timeout-to-kill-help"
                >
                  <button
                    aria-label="open-help-tooltip-button"
                    className="pf-c-form__group-label-help"
                    onClick={[Function]}
                    type="button"
                  >
                    <HelpIcon
                      color="currentColor"
                      noVerticalAlign={true}
                      size="sm"
                    />
                  </button>
                </Popover>
              }
            >
              <div />
            </mockConstructor>
          </NumberInput>
        </TimeoutToKillField>
        <PasswordField
          setValue={[Function]}
          value=""
        >
          <mockConstructor
            fieldId="job-password"
            label="Password"
            labelIcon={
              <Popover
                aria-label="help-text"
                bodyContent="Password is stored encrypted in DB until the job finishes. For future or recurring executions, it is removed after the last execution."
                id="password-help"
              >
                <button
                  aria-label="open-help-tooltip-button"
                  className="pf-c-form__group-label-help"
                  onClick={[Function]}
                  type="button"
                >
                  <HelpIcon
                    color="currentColor"
                    noVerticalAlign={true}
                    size="sm"
                  />
                </button>
              </Popover>
            }
          >
            <div />
          </mockConstructor>
        </PasswordField>
        <KeyPassphraseField
          setValue={[Function]}
          value=""
        >
          <mockConstructor
            fieldId="key-passphrase"
            label="Private key passphrase"
            labelIcon={
              <Popover
                aria-label="help-text"
                bodyContent="Key passphrase is only applicable for SSH provider. Other providers ignore this field. Passphrase is stored encrypted in DB until the job finishes. For future or recurring executions, it is removed after the last execution."
                id="key-passphrase-help"
              >
                <button
                  aria-label="open-help-tooltip-button"
                  className="pf-c-form__group-label-help"
                  onClick={[Function]}
                  type="button"
                >
                  <HelpIcon
                    color="currentColor"
                    noVerticalAlign={true}
                    size="sm"
                  />
                </button>
              </Popover>
            }
          >
            <div />
          </mockConstructor>
        </KeyPassphraseField>
        <EffectiveUserPasswordField
          setValue={[Function]}
          value=""
        >
          <mockConstructor
            fieldId="effective-user-password"
            label="Effective user password"
            labelIcon={
              <Popover
                aria-label="help-text"
                bodyContent="Effective user password is only applicable for SSH provider. Other providers ignore this field. Password is stored encrypted in DB until the job finishes. For future or recurring executions, it is removed after the last execution."
                id="effective-user-password-help"
              >
                <button
                  aria-label="open-help-tooltip-button"
                  className="pf-c-form__group-label-help"
                  onClick={[Function]}
                  type="button"
                >
                  <HelpIcon
                    color="currentColor"
                    noVerticalAlign={true}
                    size="sm"
                  />
                </button>
              </Popover>
            }
          >
            <div />
          </mockConstructor>
        </EffectiveUserPasswordField>
        <ConcurrencyLevelField
          setValue={[Function]}
          value=""
        >
          <NumberInput
            formProps={
              Object {
                "fieldId": "concurrency-level",
                "label": "Concurrency level",
                "labelIcon": <Popover
                  aria-label="help-text"
                  bodyContent="Run at most N tasks at a time. If this is set and proxy batch triggering is enabled, then tasks are triggered on the smart proxy in batches of size 1."
                  id="concurrency-level-help"
                >
                  <button
                    aria-label="open-help-tooltip-button"
                    className="pf-c-form__group-label-help"
                    onClick={[Function]}
                    type="button"
                  >
                    <HelpIcon
                      color="currentColor"
                      noVerticalAlign={true}
                      size="sm"
                    />
                  </button>
                </Popover>,
              }
            }
            inputProps={
              Object {
                "autoComplete": "concurrency-level",
                "id": "concurrency-level",
                "min": 1,
                "onChange": [Function],
                "placeholder": "For example: 1, 2, 3, 4, 5...",
                "value": "",
              }
            }
          >
            <mockConstructor
              fieldId="concurrency-level"
              helperTextInvalid="Has to be a number"
              label="Concurrency level"
              labelIcon={
                <Popover
                  aria-label="help-text"
                  bodyContent="Run at most N tasks at a time. If this is set and proxy batch triggering is enabled, then tasks are triggered on the smart proxy in batches of size 1."
                  id="concurrency-level-help"
                >
                  <button
                    aria-label="open-help-tooltip-button"
                    className="pf-c-form__group-label-help"
                    onClick={[Function]}
                    type="button"
                  >
                    <HelpIcon
                      color="currentColor"
                      noVerticalAlign={true}
                      size="sm"
                    />
                  </button>
                </Popover>
              }
            >
              <div />
            </mockConstructor>
          </NumberInput>
        </ConcurrencyLevelField>
        <TimeSpanLevelField
          setValue={[Function]}
          value=""
        >
          <NumberInput
            formProps={
              Object {
                "fieldId": "time-span",
                "label": "Time span",
                "labelIcon": <Popover
                  aria-label="help-text"
                  bodyContent="Distribute execution over N seconds. If this is set and proxy batch triggering is enabled, then tasks are triggered on the smart proxy in batches of size 1."
                  id="time-span-help"
                >
                  <button
                    aria-label="open-help-tooltip-button"
                    className="pf-c-form__group-label-help"
                    onClick={[Function]}
                    type="button"
                  >
                    <HelpIcon
                      color="currentColor"
                      noVerticalAlign={true}
                      size="sm"
                    />
                  </button>
                </Popover>,
              }
            }
            inputProps={
              Object {
                "autoComplete": "time-span",
                "id": "time-span",
                "min": 1,
                "onChange": [Function],
                "placeholder": "For example: 1, 2, 3, 4, 5...",
                "value": "",
              }
            }
          >
            <mockConstructor
              fieldId="time-span"
              helperTextInvalid="Has to be a number"
              label="Time span"
              labelIcon={
                <Popover
                  aria-label="help-text"
                  bodyContent="Distribute execution over N seconds. If this is set and proxy batch triggering is enabled, then tasks are triggered on the smart proxy in batches of size 1."
                  id="time-span-help"
                >
                  <button
                    aria-label="open-help-tooltip-button"
                    className="pf-c-form__group-label-help"
                    onClick={[Function]}
                    type="button"
                  >
                    <HelpIcon
                      color="currentColor"
                      noVerticalAlign={true}
                      size="sm"
                    />
                  </button>
                </Popover>
              }
            >
              <div />
            </mockConstructor>
          </NumberInput>
        </TimeSpanLevelField>
      </form>
    </Form>
  </AdvancedFields>
</Provider>
`;
